File: typescript/Structural/09-Decorator/Leaf/JobCard.ts
import { IJobCard } from "../Component/IJobCard";

const shortId = () => Math.random().toString(36).slice(2, 10);

export class JobCard implements IJobCard {
  readonly jobId = shortId();

  constructor(
    readonly title: string,
    readonly category: string,
    readonly cost: number,
  ) {}

  get description(): string {
    return `[${this.category}] ${this.title}`;
  }

  execute(): void {
    console.log(
      `⚙️  Executing: ${this.description} — N$${this.cost.toFixed(2)}`,
    );
  }
}

File: csharp/Structural/09-Decorator/Leaf/JobCard.cs
using Decorator.Component;

namespace Decorator.Leaf;

public class JobCard : IJobCard
{
    public string JobId { get; } = Guid.NewGuid().ToString()[..8];
    public string Title { get; }
    public string Category { get; }
    public decimal Cost { get; }
    public string Description => $"[{Category}] {Title}";

    public JobCard(string title, string category, decimal cost)
    {
        Title = title;
        Category = category;
        Cost = cost;
    }

    public void Execute() =>
        Console.WriteLine($"⚙️  Executing: {Description} — N${Cost:N2}");
}
